File: employee-management/src/main/java/com/prateek/employee_management/service/EmployeeService.java
package com.prateek.employee_management.service;

import com.prateek.employee_management.model.Employee;
import com.prateek.employee_management.repository.EmployeeRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class EmployeeService {

    @Autowired
    private EmployeeRepository employeeRepository;

    // Get all employees
    public List<Employee> getAllEmployees() {
        return employeeRepository.findAll();
    }

    // Get employee by ID
    public Employee getEmployeeById(Long id) {
        return employeeRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Employee not found with id: " + id));
    }

    // Create new employee
    public Employee createEmployee(Employee employee) {
        return employeeRepository.save(employee);
    }

    // Update employee
    public Employee updateEmployee(Long id, Employee employeeDetails) {
        Employee employee = getEmployeeById(id);
        employee.setFirstName(employeeDetails.getFirstName());
        employee.setLastName(employeeDetails.getLastName());
        employee.setEmail(employeeDetails.getEmail());
        employee.setDepartment(employeeDetails.getDepartment());
        employee.setPosition(employeeDetails.getPosition());
        return employeeRepository.save(employee);
    }

    // Delete employee
    public void deleteEmployee(Long id) {
        Employee employee = getEmployeeById(id);
        employeeRepository.delete(employee);
    }

    // Pagination getAll
    public Page<Employee> getAllEmployeesPagination(int page, int size){
        Pageable pageable = PageRequest.of(page, size, Sort.by("id").ascending());
        return employeeRepository.findAll(pageable);
    }

    // Search + Pagination
    public Page<Employee> searchEmployees(String keyword, int page, int size){
        Pageable pageable = PageRequest.of(page, size, Sort.by("id").ascending());
        return employeeRepository
                .findByFirstNameContainingIgnoreCaseOrLastNameContainingIgnoreCaseOrDepartmentContainingIgnoreCase(
                        keyword, keyword, keyword, pageable
                );
    }
}

File: employee-frontend/src/services/EmployeeService.js
import axios from 'axios';

const API_BASE_URL = 'http://localhost:8080/api/employees';


const EmployeeService = {

    // Get all Employees
     getAllEmployees: (keyword = '', page = 0, size = 5) => {
        return axios.get(`${API_BASE_URL}?keyword=${keyword}&page=${page}&size=${size}`);
     },

     // Get Employee by ID
     getEmployeeById: (id) => {
        return axios.get(`${API_BASE_URL}/${id}`);
     },

     // Create new Employee
     createEmployee: (employee) => {
        return axios.post(API_BASE_URL, employee);
     },

     // Update Employee
     updateEmployee: (id, employee) => {
        return axios.put(`${API_BASE_URL}/${id}`, employee);
     },

     // Delete Employees
     deleteEmployee: (id) => {
        return axios.delete(`${API_BASE_URL}/${id}`);
    }
}

export default EmployeeService;